concurrency/GitFarmDFDDClientBean.java

```diff
@@ -0,0 +1,448 @@
+            LOG.debug(String.format("%s ignoring forgotten %s in favor of previous, newer %s", this, toRemove, oldEntry));
+        } else {
+            LOG.info(String.format("%s removing forgotten instance %s", this, toRemove));
+
+            if ( peerToInstance.remove(toRemovePeer, oldEntry)) {
+                fireInstanceIsForgotten(toRemove);
+            } else {
+                removeInstanceIfNewer(toRemove);
+            }
+        }
+    }
+
+    private void fireInstanceStatusChange(DFDDInstance instance) {
+        for (InstanceStatusChangeListener listener : listeners) {
+            switch (instance.getStatus()) {
+                case HEALTHY:
+                    listener.instanceIsOK(instance);
+                    break;
+                case DEAD:
+                    listener.instanceIsDead(instance);
+                    break;
+                case LONGDEAD:
+                    listener.instanceIsLongDead(instance);
+                    break;
+                case TIS_ME:
+                default:
+                    // This would be a bug
+                    LOG.error(String.format("Can't notify listeners of status %s", instance.getStatus()));
+            }
+        }
+    }
+
+    private void fireInstanceIsForgotten(DFDDInstance instance) {
+        instance.setStatus(Status.LONGDEAD);
+        for (InstanceStatusChangeListener listener : listeners) {
+            listener.instanceIsForgotten(instance);
+        }
+    }
+
+    private void fireDieDieDie() {
+        for (InstanceStatusChangeListener listener : listeners) {
+            listener.diediedie();
+        }
+    }
+
+    @Override
+    public String toString() {
+        return String.format("GitFarmDFDDClientBean %s with %s known instances", this.thisInstance == null ? "<not running>" : this.thisInstance, peerToInstance.size());
+    }
+
+    @Override
+    public boolean isPeerAddress(String address) {
+        return knownPeers.contains(address);
+    }
+
+    private void addKnownPeer(DFDDInstance instance) {
+        String hostname = instance.getHostname();
+
+        if (!knownPeers.contains(hostname)) {
+            knownPeers.add(hostname);
+
+            try {
+                String ip = InetAddress.getByName(hostname).getHostAddress();
+                knownPeers.add(ip);
+            } catch (UnknownHostException e) {
+                LOG.warn(String.format("IP resolution failed for: %s", hostname), e);
+            }
+        }
+    }
+
+    class HeartbeatInstance extends DFDDInstance {
+        public final long heartbeatCount;
+        public final State state;
+
+        public HeartbeatInstance(InstanceRecord record) {
+            super(record.instanceId);
+
+            this.state = record.status;
+            this.heartbeatCount = record.heartbeatCount;
+        }
+
+        @Override
+        protected String getHeartBeatStr() {
+            return String.format("|%s", this.heartbeatCount);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format long: heartbeatCount. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 422. Consider formatting this data with &#39;%d&#39; instead.